### PR TITLE
Add customer token lifetime to GraphQl

### LINKED ISSUE #29922: CustomerToken type should contain token expired time
*As a developer I want to get the expired time for customer token so that when it's expired, I can revoke it and log out to avoid permission issues with the cart APIs.*

### Approved Schema
type CustomerToken {
    token: String @doc(description: "The customer token")
    expired_time: Int @doc(description: "Token expired time")
}

If this is approved, I will create a PR to update it.

### PR DESCRIPTION
**Pull request by codynguyen1**
### Description (*)

As a developer I want to get the expired time for customer token so that when it's expired, I can revoke it and log out to avoid permission issues with the cart APIs.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#29922

### Manual testing scenarios (*)

Send generateCustomerToken API with field expired_time on return.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by codynguyen1** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @codynguyen1. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @codynguyen1, here is your [Magento instance](https://i-29935-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-29935-2-4-develop.instances.magento-community.engineering/admin_fcde](https://i-29935-2-4-develop.instances.magento-community.engineering/admin_fcde)
Login: `f40c6fec` Password: `f0c094017944`
Instance will be terminated in up to 3 hours.

**Comment by codynguyen1** [comment]
@magento run Unit Tests

**Comment by codynguyen1** [comment]
@magento run all tests

**Comment by codynguyen1** [comment]
@magento run all tests

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (rogyar) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: Please, use `private` scope for class properties.

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (rogyar) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: I would not recommend delegating this responsibility to the resolver somehow. If the value in the database is bigger than the INT size (the system allows you to set it more than INT size), the resolver will return an incorrect value for the expiration time. Moreover, the system operates with the `access_token_lifetime` as with the Float type.

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
Comment (rogyar) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls]: I would suggest using `expiration_time` instead since `expired_time` might be literally translated as "time when the token has been expired" which does not reflect the actual value of this field 

```suggestion
    expiration_time: Float @doc(description: "Token lifetime (hours)")
```

**Review by rogyar — COMMENTED** [review]
Hi @codynguyen1. Thank you for your collaboration. 
Please, check my recommendations below.

**Comment by rogyar** [comment]
Hi @paliarush @cpartica. Could you approve the proposed schema, please? (including my review comments). 
Also, maybe it makes sense to add the `generated_at` field as well to the token. Expiration time in hours without generation time has no value unless the generation time is stored on the client side. Or we may use `expiration_date` field instead with the timestamp (instead of hours).

Hope it makes sense. 

Thank you

**Comment by codynguyen1** [comment]
Thanks, @rogyar, I have updated it.
Regarding the generated_at, the client would save the token with expired time like this:
storage.setItem('signin_token', token,expiration_time);
So the creation time is not really necessary.

**Comment by codynguyen1** [comment]
@magento run all tests

**Comment by codynguyen1** [comment]
@magento run all tests

**Comment by codynguyen1** [comment]
@rogyar the testing is returning problem with sources from other files, could you please help me bypass it? Thank you!

**Comment by rogyar** [comment]
Hi @codynguyen1. Thank you for your update. The failing tests are not related to the changes you've made. We will need to cover this change with an API-functional test but let's wait for approval from architects before moving forward.

@cpartica @paliarush could you take a look at the implementation when you have time, please? Thank you!

**Comment by nrkapoor** [comment]
@melnikovi @DrewML @prabhuram93 @cspruiell  Can you all look at the implementation and approve if it makes sense?

**Comment by nrkapoor** [comment]
@maghamed Can you provide feedback?

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (tuanpt-0634) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: There is a helper method to get this value. Should we use it instead?
```php
Magento\Integration\Helper\Oauth\Data::getCustomerTokenLifetime()
```

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (tuanpt-0634) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: Is it better to use `ttl` (time to live) instead of `expiration_time`, I think `expiration_time` is used for date time or timestamp value, but in our case this is just number of hours

**Comment by keharper** [comment]
@codynguyen1 Please open a PR against [devdocs](https://github.com/magento/devdocs) when this PR is merged. Add the new attribute to the `src/guides/v2.4/graphql/mutations/generate-customer-token.md` file, using the `2.4.2-develop` branch.

**Comment by codynguyen1** [comment]
@keharper sure I will do, but @tuanpt-0634 added some recommendations, do you guys have any comment or approval so I could apply them?

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (rogyar) [reply app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: @codynguyen1 it totally makes sense. Let's rename it to `ttl` then. Thank you!

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (rogyar) [reply app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: @codynguyen1 we could use the mentioned helper method. @tuanpt-0634 thank you for your suggestion

**Comment by rogyar** [comment]
Hi @codynguyen1. Please, check my comments above

**Comment by codynguyen1** [comment]
@magento run all tests

**Comment by codynguyen1** [comment]
Hi @rogyar I have updated the PR, but it's throwing issue from admin and magento health index, could you please help me with this?

**Comment by rogyar** [comment]
@magento run Magento Health Index

**Review by rogyar — APPROVED** [review]
The failing tests are not related to the current changeset

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-8433](https://jira.corp.magento.com/browse/ENGCOM-8433) has been created to process this Pull Request

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (tuanpt-0634) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: This is not needed

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (tuanpt-0634) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: This is not needed also

**Comment by codynguyen1** [comment]
@magento run all tests

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (codynguyen1) [reply app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php]: Sorry I missed this.

**Comment by codynguyen1** [comment]
@magento run all tests

**Comment by codynguyen1** [comment]
@magento run all tests

**Comment by codynguyen1** [comment]
@rogyar I'm sorry could you please help me with this functional tests issue again? Thanks!

**Comment by rogyar** [comment]
Hi @codynguyen1. In this particular case the failing tests are not related to your changeset. So, nothing to worry about

**Comment by gabrieldagama** [comment]
Hi @codynguyen1, thanks for contributing! Since we are moving to JWT tokens we won't process this PR, therefore I will be closing it.

Thanks!

**Pull request closed without merging**
